beacon-chain/db/slasherkv/pruning.go

```diff
@@ -0,0 +1,209 @@
+package slasherkv
+
+import (
+	"bytes"
+	"context"
+	"encoding/binary"
+
+	fssz "github.com/ferranbt/fastssz"
+	types "github.com/prysmaticlabs/eth2-types"
+	"github.com/prysmaticlabs/prysm/beacon-chain/core/helpers"
+	bolt "go.etcd.io/bbolt"
+)
+
+// PruneProposals prunes all proposal data older than currentEpoch - historyLength in
+// specified epoch increments. BoltDB cannot handle long-running transactions, so we instead
+// use a cursor-based mechanism to prune at pruningEpochIncrements by opening individual bolt
+// transactions for each pruning iteration.
+func (s *Store) PruneProposals(
+	ctx context.Context, currentEpoch, pruningEpochIncrements, historyLength types.Epoch,
+) error {
+	if currentEpoch < historyLength {
+		log.Debugf("Current epoch %d < history length %d, nothing to prune", currentEpoch, historyLength)
+		return nil
+	}
+	// We can prune everything less than the current epoch - history length.
+	endEpoch := currentEpoch - historyLength
+	endPruneSlot, err := helpers.StartSlot(endEpoch)
+	if err != nil {
+		return err
+	}
+	encodedEndPruneSlot := fssz.MarshalUint64([]byte{}, uint64(endPruneSlot))
+
+	// We retrieve the lowest stored slot in the proposals bucket.
+	var lowestSlot types.Slot
+	if err = s.db.View(func(tx *bolt.Tx) error {
+		proposalBkt := tx.Bucket(proposalRecordsBucket)
+		c := proposalBkt.Cursor()
+		k, _ := c.First()
+		lowestSlot = slotFromProposalKey(k)
+		return nil
+	}); err != nil {
+		return err
+	}
+
+	// If the lowest slot is greater than or equal to the end pruning slot,
+	// there is nothing to prune, so we return early.
+	if lowestSlot >= endPruneSlot {
+		log.Debugf("Lowest slot %d is >= pruning slot %d, nothing to prune", lowestSlot, endPruneSlot)
+		return nil
+	}
+
+	// We prune in increments of `pruningEpochIncrements` at a time to prevent
+	// a long-running bolt transaction which overwhelms CPU and memory.
+	// While we still have epochs to prune based on a cursor, we continue the pruning process.
+	epochAtCursor := helpers.SlotToEpoch(lowestSlot)
+	for endEpoch-epochAtCursor > 0 {
```

Comment:
Sorry for nitpicking, but can we change it to
```suggestion
	for epochAtCursor < endEpoch {
```
it will have the same meaning but reads a little better: like we increment `epochAtCursor`, up until it reaches the `endEpoch`.